runtime/compiler/optimizer/VectorAPIExpansion.cpp

```diff
@@ -704,6 +709,28 @@ TR_VectorAPIExpansion::expandVectorAPI()
       TR::Node *parent = NULL;
       TR::MethodSymbol *methodSymbol = NULL;
 
+
+      // remove unnecessary checkcast and NULLCHK as soon as possible
+      if (node->getOpCodeValue() == TR::checkcast ||
+         node->getOpCodeValue() == TR::NULLCHK)
+         {
+         TR::Node *firstChild = node->getFirstChild();
+         if (firstChild->getOpCodeValue() == TR::PassThrough)
+            firstChild = firstChild->getFirstChild();
+
+         if (firstChild->getOpCode().hasSymbolReference())
+            {
+            int32_t childClassId = _aliasTable[firstChild->getSymbolReference()->getReferenceNumber()]._classId;
+            if (childClassId > 0)
```

Comment:
This test is insufficient for `checkcast` because we also need to know the cast type. For example, if the `checkcast` is casting to `String`, then it should fail, but this logic would remove it

Reply:
It will only remove it it is applied to a vector object and it was vectorized or scalarized. The way library is implemented that type of object cannot cause any exception. Basically, we already proved that no checkcast is needed.

Reply:
This change ignores `checkcast` in `visitNodeToBuildVectorAliases()`, so I don't see how anything about the `checkcast` has been proved

Reply:
Yeah, it's indirectly proved. Say if we have VectorSupport.load() and then VectorSupport.binaryOp() using result of the load and we proved that all vectors produced and consumed are of the same type and length and they are not used anywhere else, we know that we can stack allocate those objects. Then, checkast on those stack temps is not needed. Basically, we know the intent of the library: that code was simply meant to add two vectors.

Reply:
Hm, while writing it I realized that the chekcast could've been from user code... Alright, let me think a bit more about it...